dirty_cat/_fuzzy_join.py

```diff
@@ -187,89 +317,125 @@ def fuzzy_join(
-            "You are merging on missing values."
-            " The output correspondence will be random or missing."
-            " To avoid unexpected errors you can drop them.",
+            "You are merging on missing values. "
+            "The output correspondence will be random or missing. "
+            "To avoid unexpected errors you can drop them. ",
             UserWarning,
         )
 
-    all_cats = pd.concat([main_col_clean, aux_col_clean], axis=0).unique()
-
-    if encoder is None:
-        enc = HashingVectorizer(analyzer=analyzer, ngram_range=ngram_range)
-    else:
-        enc = encoder
-
-    enc_cv = enc.fit(all_cats)
-    main_enc = enc_cv.transform(main_col_clean)
-    aux_enc = enc_cv.transform(aux_col_clean)
+    main_num_cols = main_table[main_cols].select_dtypes(include="number").columns
+    aux_num_cols = aux_table[aux_cols].select_dtypes(include="number").columns
 
-    all_enc = vstack((main_enc, aux_enc))
+    # Check if included columns are numeric:
+    any_numeric = len(main_num_cols) != 0
+    # Check if included columns have mixed types:
+    mixed_types = len(main_cols) > len(main_num_cols)
 
-    tfidf = TfidfTransformer().fit(all_enc)
-    main_enc = tfidf.transform(main_enc)
-    aux_enc = tfidf.transform(aux_enc)
+    if len(main_cols) == 1 and len(aux_cols) == 1 and any_numeric is False:
+        main_cols = main_cols[0]
+        aux_cols = aux_cols[0]
 
-    # Find nearest neighbor using KNN :
-    neigh = NearestNeighbors(n_neighbors=1)
-
-    neigh.fit(aux_enc)
-    distance, neighbors = neigh.kneighbors(main_enc, return_distance=True)
-    idx_closest = np.ravel(neighbors)
+    if numerical_match in ["error"] and any_numeric:
+        raise ValueError(
+            "The columns you are trying to merge on are of numerical type. "
+            "Specify numerical_match as 'string' "
+            "or 'number'. "
+        )
+    elif numerical_match in ["number"] and any_numeric and not mixed_types:
+        main_enc, aux_enc = _numeric_encoding(
+            main_table, main_num_cols, aux_table, aux_num_cols
+        )
+        idx_closest, norm_distance = _nearest_matches(main_enc, aux_enc)
+    elif numerical_match in ["number"] and any_numeric and mixed_types:
+        main_num_enc, aux_num_enc = _numeric_encoding(
+            main_table, main_num_cols, aux_table, aux_num_cols
+        )
+        main_str_cols = list(set(main_cols) - set(main_num_cols))
+        aux_str_cols = list(set(aux_cols) - set(aux_num_cols))
+        main_str_enc, aux_str_enc = _string_encoding(
+            main_table,
+            main_str_cols,
+            aux_table,
+            aux_str_cols,
+            encoder=encoder,
+            analyzer=analyzer,
+            ngram_range=ngram_range,
+        )
+        main_enc = hstack((main_num_enc, main_str_enc), format="csr")
+        aux_enc = hstack((aux_num_enc, aux_str_enc), format="csr")
+        idx_closest, norm_distance = _nearest_matches(main_enc, aux_enc)
```

Comment:
Following my comment above, I've just realized that we stack the mixed encodings together before computing the nearest neighbors. Isn't this what we want to avoid since the string encoding will be overrepresented by its number of columns?

Reply:
Yes, you are right.. 
We need to find some benchmarks for this and see what is the best solution. Do you mind keeping it this way for now or would you rather sum the distances to have an equivalent weight ?